zcm/julia/ZCM.jl

```diff
@@ -146,55 +136,55 @@ function strerror(zcm::Zcm)
+
+Adds a subscription using ZCM object `zcm` on the given channel. The `handler`
+must be a function or callable object, and will be called with:
+
+    handler(rbuf::RecvBuf, channel::String, msgdata::Vector{UInt8})
+
+If additional arguments are supplied to `subscribe()` after the handler,
+then they will also be passed to the handler each time it is called. So:
+
+    subscribe(zcm, channel, handler, X, Y, Z)
+
+will cause `handler()` to be invoked with:
+
+    handler(rbuf, channel, msgdata, X, Y, Z)
+"""
+function subscribe(zcm::Zcm, channel::AbstractString, handler,
+                   additional_args...)
+    callback = (rbuf, channel, msgdata) -> handler(rbuf, channel, msgdata, additional_args...)
+    c_handler = cfunction(handler_wrapper, Void, (Ref{Native.RecvBuf}, Cstring, Ref{typeof(callback)}))
     uv_wrapper = ccall(("uv_zcm_msg_handler_create", "libzcmjulia"),
                        Ptr{Native.UvSub},
                        (Ptr{Void}, Ptr{Void}),
-                       c_handler, Ref(options))
+                       c_handler, Ref(callback))
     uv_handler = cglobal(("uv_zcm_msg_handler_trigger", "libzcmjulia"))
     csub = ccall(("zcm_subscribe", "libzcm"), Ptr{Native.Sub},
                 (Ptr{Native.Zcm}, Cstring, Ptr{Void}, Ptr{Native.UvSub}),
                 zcm, channel, uv_handler, uv_wrapper)
-    sub = Subscription(options, c_handler, uv_wrapper, uv_handler, csub)
+    sub = Subscription(callback, c_handler, uv_wrapper, uv_handler, csub)
     push!(zcm.subscriptions, sub)
     return sub
 end
 
 """
-    subscribe(zcm::Zcm, channel::AbstractString, msgtype::Type, handler, additional_args...)
-
-Adds a subscription using ZCM object `zcm` for messages of type `msgtype` on
-the given channel. The `handler` must be a function or callable object, and
-will be called with:
-
-    handler(channel::String, msg::msgtype)
-
-If additional arguments are supplied to `subscribe()` after the handler,
-then they will also be passed to the handler each time it is called. So:
-
-    subscribe(zcm, channel, msgtype, handler, X, Y, Z)
-
-will cause `handler()` to be invoked with:
+    typed_handler(handler, msgtype)
 
-    handler(channel, msg, X, Y, Z)
+Returns a new handler function that automatically decodes the message
+data as the given msgtype before calling the original handler function
 """
-function subscribe{T <: AbstractZCMType}(zcm::Zcm,
-                                         channel::AbstractString,
-                                         msgtype::Type{T},
-                                         handler, additional_args...)
-    callback = (channel, message) -> handler(channel, message, additional_args...)
-    subscribe(zcm, channel, SubscriptionOptions(msgtype, callback))
+function typed_handler{T <: AbstractZCMType}(handler, msgtype::Type{T})
+    (rbuf, channel, msgdata, args...) -> handler(rbuf, channel, decode(msgtype, msgdata), args...)
```

Comment:
Closures aren't slower than usr pointers? That's really cool. I like Julia more and more the more I learn haha

Reply:
They used to be slow in v0.4 and earlier, but now they're just as fast as any other function. The compiler can even inline them just like other functions.